palette/src/alpha/mod.rs

```diff
@@ -0,0 +1,163 @@
+use crate::Component;
+
+#[doc(hidden)]
+pub use palette_derive::Transparency;
+
+pub use self::alpha::*;
+
+mod alpha;
+
+/// A trait for color types that can have or be given transparency (alpha channel).
+///
+/// It's an interface for adding, removing and setting the transparency of a
+/// color type. The type itself doesn't need to store the transparency
+/// value, as long as it can be transformed into or wrapped in a type that
+/// does have a representation of transparency. This would typically be done
+/// by wrapping it in an [`Alpha`](struct.Alpha.html) instance.
+///
+/// # Deriving
+/// The trait is trivial enough to be automatically derived. If the color type
+/// has a field for transparency (an alpha channel), it has to be marked with
+/// `#[palette(alpha)]` to be taken into account.
+///
+/// Derived without an internal alpha channel:
+///
+/// ```
+/// use palette::Transparency;
+///
+/// #[derive(Transparency)]
+/// struct CustomColor {
+///     redness: f32,
+///     glow: f32,
+///     glitter: f32,
+/// }
+///
+/// let color = CustomColor {
+///     redness: 0.8,
+///     glow: 2.5,
+///     glitter: 1000.0
+/// };
+/// let transparent = color.with_alpha(0.3);
+///
+/// assert_eq!(transparent.alpha, 0.3);
+/// ```
+///
+/// Derived with an internal alpha channel:
+///
+/// ```
+/// use palette::Transparency;
+///
+/// #[derive(Transparency)]
+/// struct CustomColor {
+///     redness: f32,
+///     glow: f32,
+///     glitter: f32,
+///
+///     #[palette(alpha)]
+///     alpha: u8,
+/// }
+///
+/// let color = CustomColor {
+///     redness: 0.8,
+///     glow: 2.5,
+///     glitter: 1000.0,
+///     alpha: 255
+/// };
+/// let transparent = color.with_alpha(10);
+///
+/// assert_eq!(transparent.alpha, 10);
+/// ```
+pub trait Transparency<A: Component>: Sized {
+    /// The opaque color type, without any transparency.
+    ///
+    /// This is typically `Self`.
+    type Color;
+
+    /// The color type with transparency applied.
+    ///
+    /// This is typically `Alpha<Self::Color, A>`.
+    type WithAlpha: Transparency<A, Color = Self::Color, WithAlpha = Self::WithAlpha>;
+
+    /// Transforms the color into a transparent color with the provided
+    /// alpha value. If `Self` already has a transparency, that one is
+    /// overwritten.
+    ///
+    /// ```
+    /// use palette::{Srgb, Transparency};
+    ///
+    /// let color = Srgb::new(255u8, 0, 255);
+    ///
+    /// // This results in an `Alpha<Srgb<u8>, f32>`
+    /// let transparent = color.with_alpha(0.3f32);
+    /// assert_eq!(transparent.alpha, 0.3);
+    ///
+    /// // This changes the transparency to 0.8
+    /// let transparent = transparent.with_alpha(0.8f32);
+    /// assert_eq!(transparent.alpha, 0.8);
+    /// ```
+    fn with_alpha(self, alpha: A) -> Self::WithAlpha;
+
+    /// Removes the transparency from the color. If `Self::Color` has
+    /// an internal transparency field, that one should be set to
```

Comment:
```suggestion
    /// an internal transparency field, that field will be set to
```
similar change in L119